n









diff --git a/athenapk/src/dust/dust.cpp b/athenapk/src/dust/dust.cpp
index da6ec4c..1837b60 100644
--- a/athenapk/src/dust/dust.cpp
+++ b/athenapk/src/dust/dust.cpp
@@ -1180,6 +1180,13 @@ void DustUpdateDriver(parthenon::MeshData<parthenon::Real> *md,
           DustDevObj.SetupDustForEvolutionandCoolingKernel(md);
           const auto  dust_cooling_mode_ = DustObj.dust_cooling_mode_;
+          const int  dust_scalar_idx_start = DustDevObj.dust_scalar_idx_start;
+          const int num_grain_compositions = DustDevObj.num_grain_compositions;
+          const int dust_num_grains_sizes = DustDevObj.dust_num_grains_sizes;
+          const Real mbar_gm1_over_kb = DustDevObj.mbar_gm1_over_kb;
           // FJJ Machinery for recording the AGB wind mass contributions
           int agb_history_num_rbins = 2; // some small number for low-memory usage if no AGB winds
           std::vector<double> r_bin_edges;
@@ -1214,40 +1221,21 @@ void DustUpdateDriver(parthenon::MeshData<parthenon::Real> *md,
         // FJJ TODO - think about whether to include the dust comp and size indices in the execution policy. Problem is that
         // memory starts to become scarce and kokkos starts to complain
+        if(DustDevObj.agb_winds_on == 1){
         par_for(
-        DEFAULT_LOOP_PATTERN, "Dust:DustUpdateStep", DevExecSpace(), 0,
-        cons_pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
-        KOKKOS_LAMBDA(const int &b, const int &k, const int &j, const int &i) {
+              DEFAULT_LOOP_PATTERN, "Dust:DustAGBFirstUpdateStep", DevExecSpace(), 0,
+              cons_pack.GetDim(5) - 1, 0, num_grain_compositions-1, 0, dust_num_grains_sizes-1,  kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+              KOKKOS_LAMBDA(const int &b, const int &gc_i, const int &gs_i, const int &k, const int &j, const int &i) {
               auto &cons = cons_pack(b);
               auto &prim = prim_pack(b);
               auto &coords = cons_pack.GetCoords(b);
-        const Real rho = cons(IDN, k, j, i);
-        Real internal_e =
-            cons(IEN, k, j, i) - 0.5 *
-                                     (SQR(cons(IM1, k, j, i)) + SQR(cons(IM2, k, j, i)) +
-                                      SQR(cons(IM3, k, j, i))) /
-                                     rho;
-        if (mhd_enabled) {
-          internal_e -= 0.5 * (SQR(cons(IB1, k, j, i)) + SQR(cons(IB2, k, j, i)) +
-                               SQR(cons(IB3, k, j, i)));
-        }
-        internal_e /= rho;
               const auto x = coords.Xc<1>(i);
               const auto y = coords.Xc<2>(j);
               const auto z = coords.Xc<3>(k);
               const auto r = Kokkos::sqrt(x * x + y * y + z * z);
               const auto volume =  coords.CellVolume(k, j, i);
               int rbin_idx = -1;
-    if(DustDevObj.agb_winds_on == 1){
                 // first source injection split step 
                   Real total_mass_C = 0.; // Total Mass, not a density
                   Real total_mass_S = 0.; // Total Mass, not a density
@@ -1268,11 +1256,30 @@ void DustUpdateDriver(parthenon::MeshData<parthenon::Real> *md,
                     // run in  only in this split step . otherwise will add many duplicates of the stellar masses in each sub-step
                     f_a_stellar_mass[rbin_idx] += stellar_mass_this_cell;
               }
+          });
         }
-      if (DustDevObj.dust_time_integrator_int == 1){
-          DustDoUpdateStepEuler(internal_e,
-                                      b, k, j, i,
+          par_for(
+                DEFAULT_LOOP_PATTERN, "Dust:FilladotView", DevExecSpace(), 0,
+                cons_pack.GetDim(5) - 1, 0, num_grain_compositions-1, 0, dust_num_grains_sizes-1,  kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+                KOKKOS_LAMBDA(const int &b, const int &gc_i, const int &gs_i, const int &k, const int &j, const int &i) {
+                auto &cons = cons_pack(b);
+                const auto rho = cons_pack(b,IDN, k, j, i);
+                Real internal_e =
+                    cons(IEN, k, j, i) - 0.5 *
+                                            (SQR(cons(IM1, k, j, i)) + SQR(cons(IM2, k, j, i)) +
+                                              SQR(cons(IM3, k, j, i))) /
+                                            rho;
+                if (mhd_enabled) {
+                  internal_e -= 0.5 * (SQR(cons(IB1, k, j, i)) + SQR(cons(IB2, k, j, i)) +
+                                      SQR(cons(IB3, k, j, i)));
+                }
+                internal_e /= rho;
+                auto temperature = mbar_gm1_over_kb * internal_e;
+                DustFilladotView(temperature, gc_i, gs_i, b, k, j, i,
                                             cons_pack,
                                             DustDevObj,
                                             kb, jb, ib,
@@ -1280,44 +1287,59 @@ void DustUpdateDriver(parthenon::MeshData<parthenon::Real> *md,
                                             DustDevObj.Mj_new,
                                             DustDevObj.Nj_new,
                                             DustDevObj.a_dot_view);
-      }   else if (DustDevObj.dust_time_integrator_int == 2){
-          DustDoUpdateStepHeuns(internal_e,
-                                      b, k, j, i,
+              });
+             par_for(
+                DEFAULT_LOOP_PATTERN, "Dust:DustUpdateNewNumberMassBins", DevExecSpace(), 0,
+                cons_pack.GetDim(5) - 1, 0, num_grain_compositions-1, 0, dust_num_grains_sizes-1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+                KOKKOS_LAMBDA(const int &b, const int &gc_i,const int &gs_i, const int &k, const int &j, const int &i) {
+                for(int gs_j =0; gs_j < dust_num_grains_sizes; gs_j ++){
+                  DustDoUpdateWithadotArray(gc_i, gs_i, gs_j, b, k, j, i,
                                               cons_pack,
                                               DustDevObj,
                                               kb, jb, ib,
                                               dt,
                                               DustDevObj.Mj_new,
                                               DustDevObj.Nj_new,
-                                      DustDevObj.a_dot_view,
-                                      DustDevObj.heun_state_0,
-                                      DustDevObj.heun_state_1,
-                                      DustDevObj.heun_state_2
+                                              DustDevObj.a_dot_view
                                             );
                 }
+                });
+            par_for(
+                DEFAULT_LOOP_PATTERN, "Dust:DustFinalUpdateStep", DevExecSpace(), 0,
+                cons_pack.GetDim(5) - 1, 0, num_grain_compositions-1, 0, dust_num_grains_sizes-1,  kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+                KOKKOS_LAMBDA(const int &b, const int &gc_i, const int &gs_i, const int &k, const int &j, const int &i) {
+                auto &cons = cons_pack(b);
+                auto &prim = prim_pack(b);
+                auto &coords = cons_pack.GetCoords(b);
+                const auto volume =  coords.CellVolume(k, j, i);
                 // Update cons variables from Mj_new, Nj_new obtained  performing the conservative update
-      for(int gc_i = 0; gc_i < DustDevObj.num_grain_compositions; gc_i ++){
-        for(int gs_i = 0; gs_i < DustDevObj.dust_num_grains_sizes; gs_i += 1){
                 // WriteNewDust_with_adot
                 int index_into_Ni = DustDevObj.dust_scalar_idx_start + (gc_i * 2 * DustDevObj.dust_num_grains_sizes) + (2*gs_i);
                 int index_into_Mi = index_into_Ni + 1;
                 cons(index_into_Ni, k, j, i)  = std::max(DustDevObj.Nj_new(gc_i, gs_i,b, k - kb.s, j - jb.s, i - ib.s) / volume, 0.);
                 cons(index_into_Mi, k, j, i)  = std::max(DustDevObj.Mj_new(gc_i, gs_i,b, k - kb.s, j - jb.s, i - ib.s) / volume, 0.);
                 // Just triple make sure the values are updated in both cons and cons_pack FJJ remove later once verified
                 KOKKOS_ASSERT(cons(index_into_Mi, k, j, i) == cons_pack(b,index_into_Mi, k, j, i));
                 KOKKOS_ASSERT(cons(index_into_Ni, k, j, i) == cons_pack(b,index_into_Ni, k, j, i));
-        }
-      }
             if(DustDevObj.agb_winds_on == 1){
               // second source injection split step with updated cons
+                  int rbin_idx = -1;
+                  const auto x = coords.Xc<1>(i);
+                  const auto y = coords.Xc<2>(j);
+                  const auto z = coords.Xc<3>(k);
+                  const auto r = Kokkos::sqrt(x * x + y * y + z * z);
+                  const auto volume =  coords.CellVolume(k, j, i);
                   Real total_mass_C = 0.; // Total Mass, not a density
                   Real total_mass_S = 0.; // Total Mass, not a density
                   Real stellar_mass_this_cell = 0;
@@ -1325,6 +1347,12 @@ void DustUpdateDriver(parthenon::MeshData<parthenon::Real> *md,
                     cons_pack, DustDevObj, dt / 2);
                   auto f_a_dust_agb_injected_mass_s = scatter_f_agb_injected_mass_s.access();
                   auto f_a_dust_agb_injected_mass_c = scatter_f_agb_injected_mass_c.access();
+                  auto f_a_stellar_mass             = scatter_f_stellar_mass.access();
+                  for(int rbin_i = 0; rbin_i < agb_history_num_rbins; rbin_i ++){
+                      if((r > device_r_bin_edges[rbin_i])&&(r <= device_r_bin_edges[rbin_i+1]))  {
+                      rbin_idx = rbin_i;
+                      break;}
+                  }
                   if(rbin_idx != -1){
                     f_a_dust_agb_injected_mass_c[rbin_idx] += total_mass_C;
                     f_a_dust_agb_injected_mass_s[rbin_idx] += total_mass_S;
@@ -1333,15 +1361,11 @@ void DustUpdateDriver(parthenon::MeshData<parthenon::Real> *md,
             if(DustDevObj.slope_limiting == 1){
               // slope limiting if the reconstruction method is linear, like in McKinnon
               // int gb_i = (gc_i*grain_midbin_sizes_microm.extent(0)) + gs_i; 
-      for(int gc_i = 0; gc_i < DustDevObj.num_grain_compositions; gc_i ++){
-        for(int gs_i = 0; gs_i < DustDevObj.dust_num_grains_sizes; gs_i += 1){
                 // WriteNewDust_with_adot
                 int index_into_Ni = DustDevObj.dust_scalar_idx_start + (gc_i * 2 * DustDevObj.dust_num_grains_sizes) + (2*gs_i);
                 int index_into_Mi = index_into_Ni + 1;
                 DustSlopeLimitingLinSlope(index_into_Mi, index_into_Ni,DustDevObj.code_to_microm, gs_i, gc_i, volume, cons, k, j, i, DustDevObj.grainsize_bin_edges_microm, DustDevObj.grain_midbin_sizes_microm, DustDevObj.single_grain_densities, 0); 
                 DustSlopeLimitingLinSlope(index_into_Mi, index_into_Ni,DustDevObj.code_to_microm, gs_i, gc_i, volume, cons, k, j, i, DustDevObj.grainsize_bin_edges_microm, DustDevObj.grain_midbin_sizes_microm, DustDevObj.single_grain_densities, 1);
-      }
-    }
                 } // slope_limiting == 1
               });
@@ -1358,7 +1382,7 @@ void DustUpdateDriver(parthenon::MeshData<parthenon::Real> *md,
           DustObj.WriteAGBInjectionHistory(agb_history_num_rbins, host_reduction_view_agb_injected_mass_c, host_reduction_view_agb_injected_mass_s, host_reduction_view_stellar_mass, host_r_bin_edges, md, dt, tm.time); // FJJ TODO - work out how to pass an absolue tm.time here
           }
-}
+} // void DustUpdateDriver
n









diff --git a/athenapk/src/dust/dust.hpp b/athenapk/src/dust/dust.hpp
index a3e5d18..14d8591 100644
--- a/athenapk/src/dust/dust.hpp
+++ b/athenapk/src/dust/dust.hpp
@@ -1076,6 +1076,8 @@ void SetupDustDevice(parthenon::StateDescriptor *hydro_pkg, MeshBlock *pmb){
     dust_time_integrator_int = 1;
   } else if (dust_time_integrator == "heun"){
     dust_time_integrator_int = 2;
+  } else {
+    dust_time_integrator_int = -1;
   }
   dust_num_grains_sizes   = hydro_pkg->Param<int>("dust_num_grains_sizes");
@@ -1794,10 +1797,10 @@ void GetUpdated_MjNj_ThisCompositionHelper(
 KOKKOS_INLINE_FUNCTION
-void DustDoUpdateStepEuler(
+void DustDoUpdateStepEulerInCoolingSubcycle(
     const Real internal_e,
     const int b, const int k, const int j, const int i, 
     const parthenon::MeshBlockPack<VariablePack<Real>> &cons_pack, 
@@ -1830,12 +1833,143 @@ void DustDoUpdateStepEuler(
                                           Nj_new,
                                           a_dot_view);
             } // int gc_i = 0; gc_i < num_grain_compositions; gc_i ++)
+          }
+KOKKOS_INLINE_FUNCTION
+void DustFilladotView(
+    const Real temperature,
+    const int gc_i, const int gs_i,
+    const int b, const int k, const int j, const int i, 
+    const parthenon::MeshBlockPack<VariablePack<Real>> &cons_pack, 
+    const DustDevice &DustDevObj,
+    const IndexRange &kb,
+    const IndexRange &jb,
+    const IndexRange &ib,
+    const Real sub_dt,
+    const View6DReal Mj_new,
+    const View6DReal Nj_new,
+    const View6DReal a_dot_view
+  ){
+              const int  dust_scalar_idx_start = DustDevObj.dust_scalar_idx_start;
+              const int num_grain_compositions = DustDevObj.num_grain_compositions;
+              const int dust_num_grains_sizes = DustDevObj.dust_num_grains_sizes;
+              const Real mbar_gm1_over_kb = DustDevObj.mbar_gm1_over_kb;
+                Mj_new(gc_i,  gs_i, b, k - kb.s, j - jb.s, i - ib.s) = 0.0;
+                Nj_new(gc_i,  gs_i, b, k - kb.s, j - jb.s, i - ib.s) = 0.0;
+                a_dot_view(gc_i, gs_i, b, k - kb.s, j - jb.s, i - ib.s) = 0.0;
+                Real adot_sputter = 0.;
+                Real adot_accretion = 0.;
+                Real adot = 0.;
+                const auto rho = cons_pack(b,IDN, k, j, i);
+                DustCalculateAdotPerBin(temperature, rho, DustDevObj,adot_sputter,adot_accretion,adot);
+                const Real whole_box_extent = DustDevObj.whole_box_extent;
+                const auto coords = cons_pack.GetCoords(b);
+                const auto x = coords.Xc<1>(i);
+                const auto y = coords.Xc<2>(j);
+                const auto z = coords.Xc<3>(k);
+                const auto r = Kokkos::sqrt(x * x + y * y + z * z);
+                  if(adot_sputter > 0 || adot_sputter != adot_sputter){
+                    if(std::abs(x) < 0.9*whole_box_extent && std::abs(y) < 0.9*whole_box_extent && std::abs(z) < 0.9*whole_box_extent){
                     }
+                  adot_sputter = 0.;
+                  adot_accretion = 0; // don;t do any dust updates if sputtering already is bad
+                  adot = 0.;
+                  if(std::abs(x) < 0.9*whole_box_extent && std::abs(y) < 0.9*whole_box_extent && std::abs(z) < 0.9*whole_box_extent){ // ignore weird things at box boundary e.g. negative densities
+                    PARTHENON_REQUIRE(adot_sputter <= 0, "Sputtering is growing grains!");
+                    }
+                  }
+                  if(adot_accretion < 0 || adot_accretion != adot_accretion){
+                    if(std::abs(x) < 0.9*whole_box_extent && std::abs(y) < 0.9*whole_box_extent && std::abs(z) < 0.9*whole_box_extent){
+                    }
+                  if(std::abs(x) < 0.9*whole_box_extent && std::abs(y) < 0.9*whole_box_extent && std::abs(z) < 0.9*whole_box_extent){ // ignore weird things at box boundary e.g. negative densities
+                  PARTHENON_REQUIRE(adot_accretion >= 0, "Accretion is shrinking grains!");
+                  }
+                  adot_sputter = 0.;
+                  adot_accretion = 0; // don;t do any dust updates if sputtering already is bad
+                  adot = 0.;
+                }
+                  PARTHENON_REQUIRE(a_dot_view(gc_i, gs_i, b, k - kb.s, j - jb.s, i - ib.s) == 0.0, "Contributing a_dot_view to dirty array!")
+                  a_dot_view(gc_i, gs_i, b, k - kb.s, j - jb.s, i - ib.s) = adot;
+}
+KOKKOS_INLINE_FUNCTION
+void DustDoUpdateWithadotArray(
+    const int gc_i, const int gs_i, const int gs_j,
+    const int b, const int k, const int j, const int i, 
+    const parthenon::MeshBlockPack<VariablePack<Real>> &cons_pack, 
+    const DustDevice &DustDevObj,
+    const IndexRange &kb,
+    const IndexRange &jb,
+    const IndexRange &ib,
+    const Real dt,
+    const View6DReal Mj_new,
+    const View6DReal Nj_new,
+    const View6DReal a_dot_view
+  ){
+    const int  dust_scalar_idx_start = DustDevObj.dust_scalar_idx_start;
+    const int num_grain_compositions = DustDevObj.num_grain_compositions;
+    const int dust_num_grains_sizes = DustDevObj.dust_num_grains_sizes;
+    const Real mbar_gm1_over_kb = DustDevObj.mbar_gm1_over_kb;
+    auto adot_this_i = a_dot_view(gc_i, gs_i, b, k - kb.s, j - jb.s, i - ib.s);
+    int bins_overlap = 0;
+    Real contributed_number = 0.0;
+    Real contributed_mass = 0.0;
+    const Real Mi_renorm_factor = 1;
+    DustGetMassAndNumberUpdates(bins_overlap, contributed_number, contributed_mass, gc_i, gs_i, gs_j, b,  k, j, i, cons_pack, adot_this_i, DustDevObj, dt, Mi_renorm_factor);   
+    if(bins_overlap == 1){
+      if(gs_j == dust_num_grains_sizes){
+      Mj_new(gc_i, dust_num_grains_sizes - 1, b, k - kb.s, j - jb.s, i - ib.s) += contributed_mass;
+      Nj_new(gc_i, dust_num_grains_sizes - 1, b, k - kb.s, j - jb.s, i - ib.s) += contributed_number;
+      } else{
+      Mj_new(gc_i, gs_j, b, k - kb.s, j - jb.s, i - ib.s) += contributed_mass;
+      Nj_new(gc_i, gs_j, b, k - kb.s, j - jb.s, i - ib.s) += contributed_number;      
+      }
+    } // if(bins_overlap == 1)
+  }
 KOKKOS_INLINE_FUNCTION
-void DustDoUpdateStepHeuns(
+void DustDoUpdateStepHeunsInCoolingSubcycle(
     const Real internal_e,
     const int b, const int k, const int j, const int i, 
     const parthenon::MeshBlockPack<VariablePack<Real>> &cons_pack, 
n









diff --git a/athenapk/src/hydro/hydro.cpp b/athenapk/src/hydro/hydro.cpp
index 68e8e59..070b07a 100755
--- a/athenapk/src/hydro/hydro.cpp
+++ b/athenapk/src/hydro/hydro.cpp
@@ -763,6 +763,11 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
    ************************************************************/
   auto dust_on = pin->GetOrAddBoolean("dust", "active", false);
   pkg->AddParam<bool>("dust_on", dust_on);
+  if(!dust_on){
+    pkg->AddParam<std::string>("dust_time_integrator", "none");
+    pkg->AddParam<bool>("dust_subcycle_with_cooling", "false");
+  }
n









diff --git a/athenapk/src/hydro/srcterms/tabular_cooling.cpp b/athenapk/src/hydro/srcterms/tabular_cooling.cpp
old mode 100755
new mode 100644
index 2c21be2..9e21e10
--- a/athenapk/src/hydro/srcterms/tabular_cooling.cpp
+++ b/athenapk/src/hydro/srcterms/tabular_cooling.cpp
@@ -360,10 +360,9 @@ void TabularCooling::SubcyclingFixedIntSrcTerm(MeshData<Real> *md, const Real dt
               DustObj.code_to_microm_,
   };
-  int dust_subcycle_with_cooling = hydro_pkg->Param<bool>("dust_subcycle_with_cooling") ? 1 : 0;
-  const auto  dust_cooling_mode_ = DustObj.dust_cooling_mode_;
   DustDevObj.SetupDustForEvolutionandCoolingKernel(md);
+  const auto  dust_cooling_mode_ = DustObj.dust_cooling_mode_;
   // FJJ Machinery for recording the AGB wind mass contributions
   int agb_history_num_rbins = 2; // some small number for low-memory usage if no AGB winds
@@ -603,7 +602,7 @@ par_for(
               }
               if (DustDevObj.dust_time_integrator_int == 1){
                   // update Nj_new and Mj_new Views without changing cons_pack
-                  DustDoUpdateStepEuler(internal_e,
+                  DustDoUpdateStepEulerInCoolingSubcycle(internal_e,
                                               b, k, j, i,
                                               cons_pack,
                                               DustDevObj,
@@ -614,7 +613,7 @@ par_for(
                                               DustDevObj.a_dot_view);
               }   else if (DustDevObj.dust_time_integrator_int == 2){
                     // update Nj_new and Mj_new Views without changing cons_pack
-                  DustDoUpdateStepHeuns(internal_e,
+                  DustDoUpdateStepHeunsInCoolingSubcycle(internal_e,
                                               b, k, j, i,
                                               cons_pack,
                                               DustDevObj,
n









diff --git a/athenapk/src/pgen/cluster.cpp b/athenapk/src/pgen/cluster.cpp
old mode 100755
new mode 100644
index 27398d7..d8d798b
--- a/athenapk/src/pgen/cluster.cpp
+++ b/athenapk/src/pgen/cluster.cpp
@@ -1457,12 +1457,12 @@ void UserWorkBeforeOutput(MeshBlock *pmb, ParameterInput *pin,
       DustObj.code_to_microm_,
   };
+  if (pkg->Param<bool>("dust_on")){
   DustDevObj.SetupDustDevice(pkg.get(), pmb);
+  }
   // const auto  dust_cooling_mode_ = DustObj.dust_cooling_mode_;
   // get derived fields
   auto &log10_radius = data->Get("log10_cell_radius").data;
   auto &entropy = data->Get("entropy").data;
